### PR TITLE
fix: issue with generating cypress.config.js in projects with type of module

### PR DESCRIPTION
**Pull request by ryanthemanuel**
- Closes UNIFY-1669

### User facing changelog

Projects that don't use typescript and have a type of `module` will now have their `cypress.config.js` file properly scaffolded 

### Additional details

Previously, projects that don't use typescript and have a type of `module` would generate a `cypress.config.js` using `require` and `module.exports` syntax which is invalid for those types of projects. The following error is generated:

```
[cy:open:dev:10207]: ▲ [WARNING] The CommonJS "module" variable is treated as a global variable in an ECMAScript module and may not work as expected
[cy:open:dev:10207]: 
[cy:open:dev:10207]:     cypress/plugins/index.js:19:0:
[cy:open:dev:10207]:       19 │ module.exports = (on, config) => {
[cy:open:dev:10207]:          ╵ ~~~~~~
[cy:open:dev:10207]: 
[cy:open:dev:10207]:   This file is considered to be an ECMAScript module because the enclosing "package.json" file sets the type of this file to "module":
[cy:open:dev:10207]: 
[cy:open:dev:10207]:     package.json:19:10:
[cy:open:dev:10207]:       19 │   "type": "module"
[cy:open:dev:10207]:          ╵           ~~~~~~~~
```

To fix this, we will need to flag projects that are set up with a type of `module`. When a project is flagged that way, we will ensure to use the `import`/`export default` syntax when generating a new `cypress.config` file. 

In the process of fixing this, I also noticed that migrating a legacy `module` type project with a plugin file runs into a problem since we try to require the plugin file in `run_require_async_child`. To fix this, we will fallback to the import syntax on error which is something we are already doing when loading in the config file itself.

### How has the user experience changed?

Before:

image

After:

image

### PR Tasks

- [X] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/test/.mocharc.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
 module.exports = {
-  spec: 'test/unit/**/*.spec.ts',
+  spec: 'test/unit/**/codegen.spec.ts',
```
Comment (ryanthemanuel) [inline packages/data-context/test/.mocharc.js:2]:
```suggestion
  spec: 'test/unit/**/*.spec.ts',
```

File: packages/config/src/ast-utils/addToCypressConfig.ts [hunk 2]
```diff
@@ -88,6 +88,7 @@ export interface AddToCypressConfigResult {
 }
 
 export interface AddTestingTypeToCypressConfigOptions {
+  isProjectECMAScript: boolean
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/config/src/ast-utils/addToCypressConfig.ts:91]: Doh good call. I was struggling with what to call this. I like this better

[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/config/src/ast-utils/addToCypressConfig.ts]: Resolved here: [`e02f2b1` (#21323)](https://github.com/cypress-io/cypress/pull/21323/commits/e02f2b103a18d0c3188566d9c135cac3abaf3cf0)

**Review by tgriesser — APPROVED** [review]

File: system-tests/projects/migration-e2e-module/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+## Migration E2E Defaults
+
+An e2e project with all defaults. We rename the `integrationFolder` and spec extension.
```
Comment (flotwig) [inline system-tests/projects/migration-e2e-module/README.md:3]:
```suggestion
## Migration E2E Module

An e2e project with ES modules. We rename the `integrationFolder` and spec extension.
```

**Review by flotwig — CHANGES_REQUESTED** [review]
Nice, one minor thing

**Review by flotwig — APPROVED** [review]

**Pull request merged**